### PR TITLE
PR: Initial package structure, CLI and interface ideas

### PR DESCRIPTION
**Pull request by dalthviz**
Continuation of #1

### CODE REVIEW

**Comment by ccordoba12** [comment]
I forgot to address your questions on PR #1 

> Should external executables need be part of the package but not the repo? For example for micromamba

That's a good idea, specially with micromamba. However, I don't know how to achieve that with the modern standards for Python packaging.

> Or maybe enable a way to download/install managers executables?

I think this is perhaps the simplest path to follow for now.

**Comment by mrclary** [comment]
I propose that the environment manager _not_ include executables, nor should it download and install executables (e.g. micromamba). I think that regardless of the Spyder installation method, the underlying mechanism for environment management should already be available.

The following table summarizes all the scenarios to consider for the scope of the environment manager.
Spyder Installation refers to how Spyder is installed. Runtime Env is Spyder's runtime environment.
External Conda and Venv are external environments anywhere on the user's system.
The content in the table shows the underlying mechanism that the manager would use to perform operations on the environment.

Spyder Installation | Runtime Env (plugins & Var Exp) | External Conda | External Venv
--- | --- | --- | ---
conda | runtime conda | runtime conda | runtime pip
pip | runtime pip | ? | runtime pip
standalone | runtime pip | runtime micromamba | runtime pip

- For a conda Spyder installation, both conda and pip are available from Spyder's runtime environment. These can be used to install plugins or (limited) packages into Spyder's runtime environment, as well as manipulate external environments.
- For a pip Spyder installation, only pip is available from Spyder's runtime environment. This can be used to install plugins or (limited packages into Spyder's runtime environment, but can only manipulate external pip environments. This would be the only scenario that _may_ warrant downloading micromamba. However, I don't think we should do this because a user that installs Spyder in a pip environment may be very unlikely to use external conda environments. If there is sufficient demand for this from users, we can re-evaluate it in the future.
- For standalone installations of Spyder, micromamba is already included so external conda environments can be manipulated. Pip and venv are not presently included in the macOS installer (are they included in the Windows installer?) but can easily be added; this will allow manipulating both external venv and Spyder's runtime environment. Tangentially, do we want to create a standalone installation of Spyder for Linux?

Given the above, I think that the environment manager should not need to provide any underlying executables (micromamba, pip), but just detect the appropriate available one already included with Spyder.

Additionally, I don't think there is any value in providing multiple conda environment mechanisms (conda, mamba, micromamba). In principle, it should not matter which manager is used since they all create conda environments. Attempting to provide these options to the user may overly complicate the UX, particularly for new users, as well as add more complexity to the environment manager plugin. If there is sufficient demand from users, we can re-evaluate in the future.

**Comment by dalthviz** [comment]
Thanks for the feedback @mrclary ! You think then we should leave for the moment only as manager options venv and something like 'conda-like-environment' as 'backends'? Under the hood we could try to use micromamba checking if is available and if not then falling back to conda as we try do with the standalone installers when activating conda envs?

I think for the Windows installer `venv` will need to be included. `pip` is also not included

**Comment by mrclary** [comment]
> Thanks for the feedback @mrclary ! You think that then we should leave for the moment only as manager options/'backends' venv and something like 'conda-like-environment' ? Maybe, under the hood, we could try to use micromamba checking if is available and if not then falling back to conda as we do with the standalone installers when activating conda envs in the case of the 'conda-like-environment' option.

That's correct. If Spyder is standalone, then micromamba should be available to the plugin. If Spyder is installed via conda, then the plugin won't find micromamba in the expected location but guaranteed to have a conda executable to fall back on. If Spyder is installed via pip, then the plugin may only be able to manage venv environments. The plugin could still use (or replicate) Spyder's mechanism for locating a conda executable on the system, in which case the plugin may have something to fall back on. But if the user installs via Spyder via pip and does not have any conda installed, then I think the user is very unlikely to want a conda environment. In any case, the plugin will not have to be concerned with installing any conda executable.
 
> I think for the Windows installer at least `venv` will need to be included. `pip` is also not included currently.

Given @ccordoba12's [comment](https://github.com/spyder-ide/spyder/issues/18807#issuecomment-1193409900), I think I may have misunderstood the scope of the Environment Manager. If the Environment Manager _only_ handles external environments, then we will still need another mechanism for installing third-party plugins into Spyder's runtime environment.
- Should this be another plugin or just another "widget" in Spyder? If included with Spyder, it seems that it will duplicate some of the functionality of the Environment Manager: it will require pip and it will install packages. If the Environment Manager is incorporated into Spyder, then it could handle everything, otherwise Spyder would have to include a plugin installer, separate from the Environment Manager.
- @ccordoba12 mentioned creating the installers from conda environments in order to permit installing into the Spyder runtime environment. I don't know if this is possible on macOS, given the difference between the Python frameworks (conda vs. pip). This is one issue I ran into when first investigating the macOS installer which resulted in going with pyenv so that I could install Python with the correct frameworks. Nevertheless, I don't think the user would be able to break the runtime environment since only particular packages would be allowed to be installed. For example, the existing scientific packages (matplotlib, pandas, etc.) and 3rd party plugins could be installed into a build environment and a frozen environment list included with the installer, even though not all packages in the environment would be included with the installer. Then the user may install into the runtime environment only those packages from the frozen environment list.
- I think that with the Environment Manager, the "Same as Spyder" environment is obviated. This would limit the demand for additional packages in Spyder's runtime environment to those required to extend Variable Explorer object recognition and third party plugins.

Perhaps we can discuss this in more depth on Thursday.

**Comment by ccordoba12** [comment]
> Tangentially, do we want to create a standalone installation of Spyder for Linux?

We could but there's no one to maintain it at the moment. So I'm afraid we'll have to postpone it until someone is willing to invest the time and effort to do it.

> You think that then we should leave for the moment only as manager options/'backends' venv and something like 'conda-like-environment' ?

I agree with @mrclary: we should provide conda and pip envs only. The fact that conda envs can be created with conda, mamba or micromamba is just an implementation detail.

> If Spyder is installed via pip, then the plugin may only be able to manage venv environments.

Although it's nice that we are thinking to offer the ability of creating venvs with pip, I'm afraid we couldn't have enough resources to maintain that support. I mean, for this we're probably going to need not just venv and pip but also additional managers like PDM, Poetry or pip-tools to ensure that the env is consistent after any operation (i.e. install, remove, update).

> we will still need another mechanism for installing third-party plugins into Spyder's runtime environment.

As I said, my idea for that is to switch to Conda/Mamba as the basis to build and distribute the installers. I think [Napari](https://github.com/napari/napari/releases/tag/v0.4.16) is doing something like that at the moment, so you could check how good is its Mac integration.

However, that's a different project from the one @dalthviz is working on here and would require a huge effort to implement it.

**Comment by ccordoba12** [comment]
Something I forgot to mention about Conda/Mamba to handle the Spyder env: @dalthviz and I met with the Quansight developer that helped Napari to switch to Conda/Mamba to handle its env and he told us that before that they were trying to do that with pip and it was a nightmare, i.e. people managed to break Napari easily.

That's why I think we should skip that and go directly to Conda/Mamba.

**Comment by mrclary** [comment]
> Something I forgot to mention about Conda/Mamba to handle the Spyder env: @dalthviz and I met with the Quansight developer that helped Napari to switch to Conda/Mamba to handle its env and he told us that before they were trying to do that with pip and it was a nightmare, i.e. people managed to break Napari easily.
> 
> That's why I think we should skip that step and go directly to Conda/Mamba.

I'll definitely look into that, then. It would certainly simplify a lot of things if we could use conda as the foundation for everything. If There is a robust way to make a standalone macOS application from a conda environment instead of a pip environment, I would definitely support the change.

File: env_manager/backends/conda_like_interface.py
Comment (mrclary) [inline env_manager/backends/conda_like_interface.py:5]: Should we use `pathlib` for this project?

File: env_manager/cli.py
Comment (ccordoba12) [inline env_manager/cli.py:75]:
```suggestion
    # Import env
```

**Review by ccordoba12 — COMMENTED** [review]
I left a tiny suggestion, otherwise looks good to me.

As a side note, tests are still not running but I guess they will once this PR is merged, right?

**Comment by dalthviz** [comment]
Regarding the tests, yes I would say so 👍🏼

**Review by ccordoba12 — APPROVED** [review]
Thanks @dalthviz!

**Comment by ccordoba12** [comment]
@mrclary, all good to merge?

**Review by mrclary — APPROVED** [review]

**Comment by mrclary** [comment]
Looks good to me. I approved the PR, so I'm sorry if you were waiting for me to respond.

**Pull request merged**
